.github/workflows/build_images.yml

```diff
@@ -0,0 +1,89 @@
+name: Build Images
+on:
+  push:
+    branches:
+      - dev
+      - main
+      - ci-*
+
+concurrency:
+  group: build-images-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+
+permissions:
+  id-token: write
+
+jobs:
+  build_images:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v5.0.0
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v5.1.0
+        with:
+          role-to-assume: ${{ secrets.AWS_IAM_ROLE }}
+          aws-region: us-east-1
+
+      - name: Login to Amazon ECR
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@v2
+
+      - name: Copy check into each service
+        run: |
+          cp docker/helpers/check_and_update_env.sh ./
+
+      - name: Build and push base image
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: docker/Dockerfile
+          target: ndif_base
+          outputs: type=registry,type=docker
+          tags: ${{ steps.login-ecr.outputs.registry }}/dev/ndif_base:${{ github.sha }}
+          cache-from: type=registry,ref=${{ steps.login-ecr.outputs.registry }}/dev/ndif_base:cache
+          cache-to: type=registry,ref=${{ steps.login-ecr.outputs.registry }}/dev/ndif_base:cache,mode=max
+
+      - name: Build and push API service image
```

Comment:
All four of these builds use the same single Dockerfile. The only difference is the args, and the target name (none for the services, ndif_base for the base). This is a feature of Docker build that lets you separate stages of a build.